iox-#280 Split hoofs CMakeLists.txt into multiple folders

andre-nguyen opened the pull request:
Signed-off-by: Andre Nguyen <nguyen.apv@gmail.com>

## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
<!-- Items in addition to the checklist below that the reviewer should look for -->
@elBoberido Hopefully this is what you meant in #280 ?

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes **TBD**

elBoberido commented:
@andre-nguyen sorry for the late reply. Most of the core developers are at vacation after the v2.0 release. I'll have a closer look at your PR next week

elBoberido commented on iceoryx_hoofs/CMakeLists.txt:
```diff
@@ -64,173 +64,32 @@ set(${PROJECT_NAME}_DIR ${CMAKE_CURRENT_LIST_DIR}/cmake
         PROJECT_PREFIX ${PREFIX}
     )
 
-    add_library(iceoryx_hoofs_testing
```
nice

elBoberido commented on iceoryx_hoofs/CMakeLists.txt:
```diff
@@ -64,173 +64,32 @@ set(${PROJECT_NAME}_DIR ${CMAKE_CURRENT_LIST_DIR}/cmake
-########## build iceoryx util lib ##########
-#
-
-add_library(iceoryx_hoofs
```
Could you please revert this?

Since the header for this target are in the `include` folder which lies alongside the `source` folder, one would need to step out of `source` to add the `target_include_directories`.

For a future refactoring one could think of moving the include folder to source, like
```
iceoryx_hoofs
|- cmake
|- platform
|- source
|   |- include
|   |    |- iceoryx_hoofs
|   |- *.cpp or a cpp folder?
|- test
|- testing
```

@dkroenke what do you think?

dkroenke replied:
@elBoberido 
It looks a bit strange when there is an `include` folder below `source` and looks a bit unintuitive.
A `cpp` folder is strange too because the include folder should be then named `hpp` to be consistent...

I understand that the CMakeLists.txt in `iceoryx_hoofs` is quite large because of certain things like iceoryx_hoofs_testing or the Clang-Tidy statements.
Nevertheless it makes more sense imho to have in the `iceoryx_hoofs` folder a CMakeLists.txt that builds the iceoryx_hoofs lib itself. Other things like the testing lib, Gtest, Clang-Tidy or the packaging commands can be outsourced in external cmake files.
With that, the size of the CMakeLists.txt file in iceoryx_hoofs will be maintainable from my point of view.

elBoberido replied:
@dkroenke yes, this does not need to be decided now. I just thought about something like this https://palikar.github.io/posts/cmake_structure/

At some point we might want to modularize the hoofs a bit more, e.g. having `posix_wrapper` and `cxx` as separate modules. Instead of having a top-level `source` folder maybe we could have a `code` or `modules` folder, e.g.
```
iceoryx_hoofs
|- cmake
|- platform
|- modules
|   |- cxx
|   |   |- include
|   |   |- source
|   |- posix_wrapper
|   |    |- ...
|- test
|- testing
```

But like already mentioned, this is something which can be done in a later PR. At first we need to find out which direction we want to go. The current separation is already an improvement from what is in master.

andre-nguyen commented:
Thanks for the review, I've addressed the comments however, in my rebase and force-push It seems I've lost the log of the previous failed pipeline so I'm not sure where the error was. Hopefully it passes now??

andre-nguyen commented:
Oh I forgot to address the alias target creation, pushed that commit now.

andre-nguyen commented:
@elBoberido Do you mind checking the failure log? It doesn't seem like the build failure is related to my changes. It is unclear to me why rsync would fail and while code coverage would get a 404.

elBoberido commented:
@andre-nguyen I re-triggered the failed jobs. Let's see what happens

andre-nguyen commented:
```
Bad packet length 2005070767.
ssh_dispatch_run_fatal: Connection to 127.0.0.1 port 2222: Connection corrupted
Error: The process '/usr/bin/ssh' failed with exit code 255
get back by rsync
/usr/bin/rsync -uvzrtopg freebsd:work/ /Users/runner/work
receiving file list ... done
```

Hhhm seems to be the [same error](https://github.com/eclipse-iceoryx/iceoryx/runs/5980030639?check_suite_focus=true#step:3:38471) as before. Any idea what this means?

dkroenke commented on iceoryx_hoofs/CMakeLists.txt:
```diff
@@ -64,86 +64,18 @@ set(${PROJECT_NAME}_DIR ${CMAKE_CURRENT_LIST_DIR}/cmake
-# TODO: we need something like internal public for internal build dependencies
-# instead of making the whole source folder public for internal dependency resolution
-
 if(BUILD_TEST AND NOT GTest_FOUND)
     find_package(GTest CONFIG REQUIRED)
 endif(BUILD_TEST AND NOT GTest_FOUND)
```
Maybe we can find a way to group all the test related CMake Code a bit more together here.
Ideally we have something like:

```cmake
########## hoofs testing ##########
if(BUILD_TEST AND NOT GTest_FOUND)
    find_package(GTest CONFIG REQUIRED)
endif(BUILD_TEST AND NOT GTest_FOUND)

add_subdirectory(testing)
add_subdirectory(test) #taken from the bottom of the file

...
other sections
```

andre-nguyen replied:
The original logic didn't make sense did it?

`iceoryx_hoofs_testing` was only created if gtest was found and a comment mentioned that you could build it without building the unit tests. However, the `find_package Gtest` call was only performed if we did build the unit tests, therefore there wasn't actually a way to build `iceoryx_hoofs_testing` without `BUILD_TESTS=ON`. Thus I merged everything together in the same if clause.

Let me know if that makes sense.

dkroenke commented:
> ```
> Bad packet length 2005070767.
> ssh_dispatch_run_fatal: Connection to 127.0.0.1 port 2222: Connection corrupted
> Error: The process '/usr/bin/ssh' failed with exit code 255
> get back by rsync
> /usr/bin/rsync -uvzrtopg freebsd:work/ /Users/runner/work
> receiving file list ... done
> ```
> 
> Hhhm seems to be the [same error](https://github.com/eclipse-iceoryx/iceoryx/runs/5980030639?check_suite_focus=true#step:3:38471) as before. Any idea what this means?

@andre-nguyen 
It was a hiccup on CI and not related to your changes, I had some final remarks, otherwise it looks good already

andre-nguyen commented on iceoryx_hoofs/CMakeLists.txt:
```diff
@@ -64,86 +64,18 @@ set(${PROJECT_NAME}_DIR ${CMAKE_CURRENT_LIST_DIR}/cmake
         NAME ${PROJECT_NAME}_testing
         NAMESPACE iceoryx_hoofs_testing
         PROJECT_PREFIX ${PREFIX}
     )
```
Done, had to modify `setup_package_name_and_create_files` macro to fetch the config file from the project root.

elBoberido commented on iceoryx_hoofs/CMakeLists.txt:
```diff
@@ -230,10 +147,20 @@ target_compile_options(iceoryx_hoofs PRIVATE ${ICEORYX_WARNINGS} ${ICEORYX_SANIT
+########## hoofs testing ##########
+#
+
+if(BUILD_TEST)
+    find_package(GTest CONFIG REQUIRED)
+    add_subdirectory(testing)
+    add_subdirectory(test)
+endif(BUILD_TEST)
```
This changes the semantics for the testing library. Before it was build when gtest was found, now only when the tests are build. Le's assume someone integrates icoryx as subrepo in a bigger code-base and wants to use the testing lib then it is not anymore possible without also building all the iceoryx tests.

@dkroenke do we want to keep being able to build the testing lib without building the tests?

andre-nguyen replied:
Just to confirm my understanding here, before we had the following snippet

``` cmake
if(BUILD_TEST AND NOT GTest_FOUND)
    find_package(GTest CONFIG REQUIRED)
endif(BUILD_TEST AND NOT GTest_FOUND)
```

which meant that gtest was only looked for if `BUILD_TEST` was `ON`. That means that `testing` was only built if the tests were being built.

Are you saying that if you for example do `add_subdirectory(iceoryx/iceoryx_hoofs)` a `find_package(GTest REQUIRED)` call from the parent project will transfer to this one and enable `testing`?

for example
```cmake
project(parent_proj)

find_package(GTest REQUIRED)

add_subdirectory(iceoryx/iceoryx_hoofs)  # <<--- Will this call enable iceoryx_hoofs_testing??
```

elBoberido replied:
@andre-nguyen yes, that would enable the `testing` lib. In case someone is building a framework on top of iceoryx, this might be desirable. One could for example use a released version of iceoryx without increasing the build time by also building the tests. I'm just not sure if we want to solve the problem this way. Maybe a dedicated cmake switch would make more sense if someone really needs use iceoryx the way I described it.

My tendency is to simplify it like you did and introduce this when someone needs it but let's see if @dkroenke agrees.

dkroenke replied:
@elBoberido @andre-nguyen 

Normally the `testing` functions are tightly coupled the the iceoryx tests. I see there generic functions like the timing test stubs but on the other hand quite iceoryx -specific mocks like a mock for the error handler.

Then we could do something like this:
```cmake
if(NOT GTest_FOUND AND BUILD_TEST)
    find_package(GTest CONFIG REQUIRED)
elseif(GTest_FOUND)
    add_subdirectory(testing)
elseif(GTest_FOUND AND BUILD_TEST)
   add_subdirectory(test)
endif()
```

This snippet is untested, this could work maybe, @andre-nguyen can you test that?
The intention is not to save codelines but have all the test-related logic collected a bit into a section in cmake.

elBoberido replied:
@dkroenke that code snippet won't work. It must be something like
```cmake
# finding gtest and adding the subdirectories is split to support the use case of
# building the testing lib without the tests by providing gtest externally
if(NOT GTest_FOUND AND BUILD_TEST)
    find_package(GTest CONFIG REQUIRED)
endif()

if(GTest_FOUND)
    add_subdirectory(testing)
    if(BUILD_TEST)
        add_subdirectory(test)
    endif()
endif()
```

elBoberido reviewed (commented):
Looks good. Just one minor thing we have to digest

elBoberido commented:
@andre-nguyen looks good now. Please rebase/merge master. This will fix the FreeBSD CI build and then this is ready from my point of view.

elBoberido reviewed (approved):
Thanks for your patience :)
